core/commonTest/src/kotlinx/serialization/SerializersModuleTest.kt

```diff
@@ -0,0 +1,126 @@
+
+    @Serializer(forClass = WithCustomSerializer::class)
+    object CustomSerializer
+
+    @Serializable
+    class Parametrized<T : Any>(val a: T)
+
+    class ContextualType(val i: Int)
+
+    class ParametrizedContextual<T : Any>(val a: T)
+
+    @Serializer(forClass = ContextualType::class)
+    object ContextualSerializer
+
+    @Serializer(forClass = ParametrizedContextual::class)
+    object ParametrizedContextualSerializer
+
+    class FileContextualType(val i: Int)
+
+    @Serializer(forClass = FileContextualType::class)
+    object FileContextualSerializer
+
+    @Serializable
+    class ContextualHolder(@Contextual val contextual: ContextualType, val fileContextual: FileContextualType)
+
+    @Test
+    fun testCompiled() = noJsLegacy {
+        assertSame<KSerializer<*>>(Object.serializer(), serializer(Object::class, emptyList(), false))
+        assertSame<KSerializer<*>>(SealedParent.serializer(), serializer(SealedParent::class, emptyList(), false))
+        assertSame<KSerializer<*>>(
+            SealedParent.Child.serializer(),
+            serializer(SealedParent.Child::class, emptyList(), false)
+        )
+
+        assertSame<KSerializer<*>>(Abstract.serializer(), serializer(Abstract::class, emptyList(), false))
+        assertSame<KSerializer<*>>(SerializableEnum.serializer(), serializer(SerializableEnum::class, emptyList(), false))
+    }
+
+    @Test
+    fun testBuiltIn() {
+        assertSame<KSerializer<*>>(Int.serializer(), serializer(Int::class, emptyList(), false))
+    }
+
+    @Test
+    fun testCustom() {
+        val m = SerializersModule { }
+        assertSame<KSerializer<*>>(CustomSerializer, m.serializer(WithCustomSerializer::class, emptyList(), false))
+    }
+
+    @Test
+    fun testParametrized() {
+        val serializer = serializer(Parametrized::class, listOf(Int.serializer()), false)
+        assertEquals<KClass<*>>(Parametrized.serializer(Int.serializer())::class, serializer::class)
+        assertEquals(PrimitiveKind.INT, serializer.descriptor.getElementDescriptor(0).kind)
+
+        val mapSerializer = serializer(Map::class, listOf(String.serializer(), Int.serializer()), false)
+        assertIs<MapLikeSerializer<*, *, *, *>>(mapSerializer)
+        assertEquals(PrimitiveKind.STRING, mapSerializer.descriptor.getElementDescriptor(0).kind)
+        assertEquals(PrimitiveKind.INT, mapSerializer.descriptor.getElementDescriptor(1).kind)
+    }
+
+    @Test
+    fun testUnsupportedArray() {
+        assertFails {
+            serializer(Array::class, listOf(Int.serializer()), false)
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    @Test
+    fun testContextual() {
+        val m = SerializersModule {
+            contextual<FileContextualType>(FileContextualSerializer)
+            contextual<ContextualType>(ContextualSerializer)
+            contextual<ParametrizedContextual<*>>(ParametrizedContextualSerializer as KSerializer<ParametrizedContextual<*>>)
```

Comment:
there's `contextual` that accepts lambda and is able to pass type arguments' serializers: https://github.com/Kotlin/kotlinx.serialization/blob/c6025c9fb6ed517169b4a5516dcd210bc2d7576a/core/commonTest/src/kotlinx/serialization/modules/ContextualGenericsTest.kt#L48